### DISCUSSION TITLE
Is there a way to add context to an instance of EntityManager?

### DISCUSSION THREAD

**Discussion by epechuzal** [mikro-orm/mikro-orm#2188] [Q&A]
**Is your feature request related to a problem? Please describe.**
I am looking for a way to add a kind of paper trail system in my app: whenever certain entities are created/updated/deleted I want to log:
- what entity was changed
- what was changed about it
- who did it

Thanks to your [docs](https://mikro-orm.io/docs/lifecycle-hooks/#flush-events) the first two points are easy to do. I am struggling with the third point.

**Describe the solution you'd like**
It is my understanding that in NestJS whenever a HTTP request is received there is a fork made of the entity manager. Is it possible to add information to this forked entity manager? Specifically, I would like it to have a copy of the `req.user` object so that I can also log who did a change.

**Describe alternatives you've considered**
I have thought of a couple ideas but I think that I do not know enough about how nest and mikro-orm work in order to know if they are feasible.

First idea: somehow add `req.user` object to the entity manager? That way we could do something like:
```ts
async onFlush(args: FlushEventArgs): Promise<void> {
    const changeSets = args.uow.getChangeSets();

    for (const cs of changeSets) {
      const auditLogEntry = new AuditLogEntry();

      auditLogEntry.entityType    = cs.entity.constructor.name;
      auditLogEntry.entityId      = cs.entity.id;
      auditLogEntry.object        = cs.originalEntity;
      auditLogEntry.objectChanges = cs.payload;
      auditLogEntry.event         = cs.type;
      // where `ext` is of type `Record<string, unknown>` and allows users to add JSON info to the EM object
      auditLogEntry.whodunnit = args.em.ext.user.email;

      args.uow.computeChangeSet(auditLogEntry)
    }
  }
```

Second idea:
Can our subscriber be itself an injectable class? then we could use it as a [request scoped provider](https://docs.nestjs.com/fundamentals/injection-scopes#request-provider)? e.g. 
```ts
@Injectable({ scope: Scope.REQUEST })
export class PaperTrailSubscriber implements EventSubscriber<AnyEntity> {
  constructor(@Inject(REQUEST) private request: Request) {}

  getSubscribedEntities(): EntityName<AnyEntity>[] {
    return [EntityA, EntityB];
  }

  async onFlush(args: FlushEventArgs): Promise<void> {
    const changeSets = args.uow.getChangeSets();

    for (const cs of changeSets) {
      const auditLogEntry = new AuditLogEntry();

      auditLogEntry.entityType    = cs.entity.constructor.name;
      auditLogEntry.entityId      = cs.entity.id;
      auditLogEntry.object        = cs.originalEntity;
      auditLogEntry.objectChanges = cs.payload;
      auditLogEntry.event         = cs.type;
      auditLogEntry.whodunnit = this.request.user.email;

      args.uow.computeChangeSet(auditLogEntry)
    }
  }
}
```

I feel like I am probably thinking about this the wrong way, so I am curious what your thoughts are on this.

**Comment by B4nan** [comment]
You could use `em.setFilterParams()` and `em.getFilterParams()` as a general storage for anything. Alternatively you could also use your own instance of `AsyncLocalStorage`.

> Can our subscriber be itself an injectable class? then we could use it as a request scoped provider? e.g.

Subscribers can be injected, but I am not sure if it would work correctly here, as they are stored globally in v4 (see https://github.com/mikro-orm/mikro-orm/pull/1741 for how it works in v5). You would have to register the subscriber in your constructor (while injecting the EM into it):

```ts
@Injectable()
export class MySubscriber {

  constructor(em: EntityManager) {
    em.getEventManager().registerSubscriber(this);
  }

}
```

**Reply by epechuzal** [reply]
So from what you are saying I feel the best bet is to go the injection route but only after v5 is released (do you have a rough idea of when?)

In the meantime, it sounds like I could write some kind of middleware that runs after authentication, where I inject the Entity Manager, call `em.setFilterParams(<User object>)`, and then in my subscriber I can just call `em.getFilterParams()` to recover the logged in user. Would this work? When injecting the entity manager, does nestjs return the forked EM from a request, or some global instance?

**Comment by epechuzal** [comment]
Actually I found a third way which works for now using interceptors
```ts
@Injectable()
export class PaperTrailInterceptor implements NestInterceptor {
  constructor(
    private readonly em: EntityManager,
  ) {
  }

  intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
    const user = context.switchToHttp().getRequest().user;
    this.em.setFilterParams('user', user);

    return next.handle();
  }
}
```

This will grab currently logged in user right before a controller is going to do some work.

Then in our subscriber we just need these changes
```ts
const user = args.em.getFilterParams('user');
auditLogEntry.whodunnit     = user?.email ?? 'unknown';
```

When you release v5 maybe the approach with injectable subscribers will work better, but this seems to work so far. What do you think?
